### PR TITLE
 #3570/broken pi-manage backup when database port is defined

### LINKED ISSUE #3570: Ruining pi-manage backup error when database port is defined
### Top-level intent

What did you try to achieve?
Create backup from `pi-manage backup create` should run successfully when the database port is defined in  ***SQLALCHEMY_DATABASE_URI***
### Steps to reproduce

1. In pi.cfg define ***SQLALCHEMY_DATABASE_URI*** with database port e.g. `SQLALCHEMY_DATABASE_URI = "pymysql://privacyidea:privacyidea@privacyidea-db:3306/privacyidea"`
2. Run `pi-manage backup create`

### Expected outcome

What do you think, how the system should have worked?
Backup should create successfully

### Actual outcome

What did actually happen?
There is an error with message `mysqldump: Got error: 2005: "Unknown MySQL server host 'privacyidea-db:3306' (-2)" when trying to connect`

### Configuration
SQLALCHEMY_DATABASE_URI = "pymysql://privacyidea:privacyidea@privacyidea-db:3306/privacyidea"

* **privacyIDEA version**: 3.8.1
* **Installation method**: PyPI (from Ubuntu packages, github, PyPI, ...)
* **Python version**: 3.8.16
* **Operating system**: Debian 11
* **Webserver**: 
* **Token database**:  MySQL (MySQL, PostgreSQL, ...)

### Log file

**Set PI_LOGLEVEL = logging.DEBUG in pi.cfg and take a look at the privacyidea.log!**
**If appropriate, attach the log file or paste relevant portions.**

### PR DESCRIPTION
**Pull request by NNLemling**
- [x] change to urllib.parse for splitting the database uri
- [x] check tests

closes #3570

### CODE REVIEW

File: pi-manage [hunk 1]
```diff
@@ -297,25 +298,26 @@ def create(directory="/var/lib/privacyidea/backup/",
...
-        password = m.groups()[1]
-        datahost = m.groups()[2]
-        database = m.groups()[3]
+        username = parsed_sqluri.username
+        password = parsed_sqluri.password
+        hostname = parsed_sqluri.hostname
+        database = parsed_sqluri.path[1:]
         # We strip parameters, but we do not use them
-        _parameters = m.groups()[4]
+        _parameters = parsed_sqluri.params
```
Comment (plettich) [inline pi-manage:315]: I think we can remove these 2 lines if they are not used

File: pi-manage [hunk 2]
```diff
@@ -297,25 +298,26 @@ def create(directory="/var/lib/privacyidea/backup/",
...
         defaults_file = "{0!s}/mysql.cnf".format(conf_dir)
         _write_mysql_defaults(defaults_file, username, password)
         call("mysqldump --defaults-file=%s -h %s %s > %s" % (
             shlex_quote(defaults_file),
-            shlex_quote(datahost),
+            shlex_quote(hostname),
```
Comment (plettich) [inline pi-manage:320]: The port information is missing here. A better approach would be something like:
```
cmd = ['mysqldump', '--defaults-file={!s}'.format(shlex_quote(defaults_file)), '-h', shlex_quote(hostname)]
if parsed_sqluri.port:
    cmd.extend(['-p', parsed_sqluri.port])
cmd.extend(['-B', shlex_quote(database), '-r', shlex_quote(sqlfile)])
call(cmd, shell=True)
```

**Review by plettich — CHANGES_REQUESTED** [review]

File: pi-manage (hunk 1 above)
Comment (NNLemling) [reply pi-manage]: fixed in 71167a5072fe3f5e7ef36903d2e68649ab66a5a7.

File: pi-manage (hunk 2 above)
Comment (NNLemling) [reply pi-manage]: fixed in 71167a5072fe3f5e7ef36903d2e68649ab66a5a7.

File: pi-manage [hunk 3]
```diff
@@ -297,27 +298,28 @@ def create(directory="/var/lib/privacyidea/backup/",
...
-        call("mysqldump --defaults-file=%s -h %s %s > %s" % (
-            shlex_quote(defaults_file),
-            shlex_quote(datahost),
-            shlex_quote(database),
-            shlex_quote(sqlfile)), shell=True)
+        cmd = ['mysqldump', '--defaults-file={!s}'.format(
+            shlex_quote(defaults_file)), '-h',
+            shlex_quote(hostname)]
+        if parsed_sqluri.port:
+            cmd.extend(['-p', parsed_sqluri.port])
```
Comment (plettich) [inline pi-manage:320]: The port number needs to be a string as well.
And somehow `mysqldump` doesn't work with these parameters. I'll have to check.

Comment (NNLemling) [reply pi-manage]: fixed in cc453db88a442d312081d142089401aaf81a870a.

File: pi-manage [hunk 4]
```diff
@@ -297,27 +298,28 @@ def create(directory="/var/lib/privacyidea/backup/",
...
-        call("mysqldump --defaults-file=%s -h %s %s > %s" % (
-            shlex_quote(defaults_file),
-            shlex_quote(datahost),
-            shlex_quote(database),
-            shlex_quote(sqlfile)), shell=True)
+        cmd = ['mysqldump', '--defaults-file={!s}'.format(
+            shlex_quote(defaults_file)), '-h',
+            shlex_quote(hostname)]
+        if parsed_sqluri.port:
+            cmd.extend(['-p', str(parsed_sqluri.port)])
```
Comment (plettich) [inline pi-manage:320]: The command-line switch for "port" is `-P` (capital "P"). That is why it didn't work. My bad.

**Review by plettich — CHANGES_REQUESTED** [review]

File: pi-manage (hunk 4 above)
Comment (NNLemling) [reply pi-manage]: fixed in 1cb4bc4ecdae34f0d8f4b33df25fc95bae9e59f9.

**Review by plettich — APPROVED** [review]

**Pull request merged**